src/library/scala/collection/immutable/ListMap.scala

```diff
@@ -127,10 +127,18 @@ sealed class ListMap[K, +V]
       else containsInternal(cur.next, k)
 
     override def updated[V2 >: V1](k: K, v: V2): ListMap[K, V2] = {
-      val (m, k0) = removeInternal(k, this, Nil)
-      new m.Node[V2](k0, v)
+      replaceInternal(k, v, this, Nil)
     }
 
+    @tailrec private[this] def replaceInternal[V2 >: V1](k: K, v : V2, cur: ListMap[K, V2], acc: List[ListMap[K, V2]]): ListMap[K, V2] =
+      if (cur.isEmpty) new this.Node(k,v)
+      else if (k == cur.key) {
+        val nextNode : ListMap[K, V2] = cur.next
+        val updatedCurrentNode : ListMap[K, V2] = new nextNode.Node(cur.key,v)
```

Comment:
You could check if `cur.value eq v` and simply return `this` in this case instead of unnecessarily finishing the transformation.